finesse/hardware/st10_controller.py

```diff
@@ -19,6 +22,113 @@ class ST10ControllerError(SerialException):
     """Indicates that an error has occurred with the ST10 controller."""
 
 
+_ASYNC_MAGIC = "Z"
+
+
+class _SerialReader(QThread):
+    """For background reading of serial device."""
+
+    async_read_completed = Signal()
+    """Indicates that an asynchronous read has finished."""
+
+    read_error = Signal(BaseException)
+    """Sent when an error occurs during reading."""
+
+    def __init__(self, serial: Serial, sync_timeout: float) -> None:
+        """Create a new _SerialReader.
+
+        Args:
+            serial: Serial device
+            sync_timeout: Read timeout for synchronous requests
+        """
+        super().__init__()
+
+        self.serial = serial
+        self.sync_timeout = sync_timeout
+        self.out_queue: Queue = Queue()
+        self.stopping = False
+
+    def __del__(self) -> None:
+        """Wait for the thread to stop on exit."""
+        self.quit()
+        self.wait()
+
+    def quit(self) -> None:
+        """Flag that the thread is stopping so we can ignore exceptions."""
+        self.stopping = True
+        super().quit()
+
+    def _read(self) -> str:
+        """Read the next message from the device.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        raw = self.serial.read_until(b"\r")
+
+        # Check that it hasn't timed out
+        if not raw:
+            raise SerialTimeoutException()
+
+        logging.debug(f"(ST10) <<< {repr(raw)}")
+
+        try:
+            return raw[:-1].decode("ascii")
+        except UnicodeDecodeError:
+            raise ST10ControllerError(f"Invalid message received: {repr(raw)}")
+
+    def _read_error(self, error: BaseException) -> None:
+        # Return the error synchronously to the first waiter
+        self.out_queue.put(error)
+
+        # Also send error as a signal as there is not necessarily a synchronous waiter
+        self.read_error.emit(error)
+
+    def _read_success(self, message: str) -> None:
+        # The motor is signalling that it has finished moving
+        if message == _ASYNC_MAGIC:
+            self.async_read_completed.emit()
+            return
```

Comment:
The only asynchronous message that we'll be receiving is the magic string we've requested (i.e. "Z"). All other messages *should* be ones that have been requested explicitly. We need to keep track of the other messages that were received because there might be a synchronous waiter (which will be running on a different thread).

For example, this situation might happen:

1. Main thread: The code sends a synchronous message, e.g. it asks the controller the position of the motor
2. Reader thread: The `_SerialReader` receives the response saying what the position is
3. Reader thread: It also happens to receive the magic "Z" message just after this (and before the main thread has had a chance to process the last message received)
4. Main thread: Retrieve the message from the reader's queue

So if we emptied the queue on receiving the Z message, then there wouldn't be anything left for the main thread to read. The whole async/send string thing is really separate from the normal synchronous messages we're sending.